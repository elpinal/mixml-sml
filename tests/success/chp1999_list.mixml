link L = {unit LIST =
  link X = {type t : Type} with
  link {val nil : ext X.t} with
  link {val null : ext X.t -> bool} with
  link {val cons : int -> ext X.t -> ext X.t} with
  {val uncons : ext X.t -> int * ext X.t}
} with

{module List =
  ; forward declaration
  link X = new L.LIST with

  ; shorthand
  link S = {type u = [Nil : unit, Cons : int * ext X.t]} with

  ; the list type
  link I = new [unit
    link
      link T = {type t : Type} with
      link {val t_in : ext S.u -> ext T.t} with
      {val t_out : ext T.t -> ext S.u}
    seals
    link T = {type t = ext S.u} with
    link {val t_in = fun (x : ext T.t) -> x} with
    {val t_out = fun (x : ext T.t) -> x}
  ] with

  link {val nil = ext I.t_in (Nil () : ext S.u)} with
  link {val null = fun (xs : ext I.t) ->
    match ext I.t_out xs => {
      | Nil () => true
      | Cons _ => false
    }
  } with
  link {val cons = fun (n : int) (l : ext I.t) -> ext I.t_in (Cons (n, l) : ext S.u)} with
  {val uncons = fun (l : ext I.t) ->
    match ext I.t_out l => {
      | Nil ()      => (0, l)
      | Cons (n, l) => (n, l)
    }
  }
}
