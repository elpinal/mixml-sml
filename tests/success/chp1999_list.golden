{LIST : [forall (!476@t : Type). {cons : [int -> !476 -> !476]-, nil : [!476]-, null : [!476 -> bool]-, t : [= !476], uncons : [!476 -> int * !476]-}]+, List : {cons : [int -> !0 -> !0]+, nil : [!0]+, null : [!0 -> bool]+, t : [= !0], t_in : [[Cons : int * !0, Nil : unit] -> !0]+, t_out : [!0 -> [Cons : int * !0, Nil : unit]]+, u : [= [Cons : int * !0, Nil : unit]], uncons : [!0 -> int * !0]+}}
